### DISCUSSION TITLE
I'm getting BAD_REQUEST after deploy on vercel

### DISCUSSION THREAD

**Discussion by MunnyReol** [vercel/next.js#43475] [Help]
### What version of Next.js are you using?

10.1.3

### What version of Node.js are you using?

14.15.4

### What browser are you using?

Chrome

### What operating system are you using?

Windows

### How are you deploying your application?

Vercel

### Describe the Bug

I want to take screenshot of webpage using next js that's why i created a dynamic api.
but when i request to website i get the 404 erro & BAD_REQUEST.

My screenshot path is `pages/api/screenshot/[id]/index.js`

My code is like this
```
async function api(req, res) {    
    res.send(req.query.id)
}

export default api
```

When i request to website i get 404 the url is like this [https://next-dynamic-munnyreol.vercel.app/api/screenshot/https%3A%2F%2Fnextjs.org](https://next-dynamic-munnyreol.vercel.app/api/screenshot/https%3A%2F%2Fnextjs.org) 

I add (encodeURIComponent) url after slash screenshot
 `https%3A%2F%2Fnextjs.org` mean `https://nextjs.org`
.
.
I'm getting BAD_REQUEST when i request to [https://next-dynamic-munnyreol.vercel.app/api/screenshot/https%3A%2F%2](https://next-dynamic-munnyreol.vercel.app/api/screenshot/https%3A%2F%2) 

### Expected Behavior

When i request to `https://next-dynamic-munnyreol.vercel.app/api/screenshot/https%3A%2F%2Fnextjs.org` then i want to get the value which i put after slash screenshot `screenshot/`
For this request, i want to get `https://nextjs.org` as a output

### To Reproduce

The GitHub repository is 
[https://github.com/MunnyReol/next-dynamic](https://github.com/MunnyReol/next-dynamic)

**Comment by ljosberinn** [comment]
This is not Next.js related but with how URLs work. Pass the url instead as GET parameter, e.g.
```js
https://next-dynamic-munnyreol.vercel.app/api/screenshot?url=https%3A%2F%2Fnextjs.org
```
```js
export default function handler(req, res) {
  if(!req.param.url || Array.isArray(req.param.url)) {
    res.status(400).end();
    return;
  }
  
  res.send(req.param.url);
}
```

**Comment by MunnyReol** [comment]
> This is not Next.js related but with how URLs work. Pass the url instead as GET parameter, e.g.
> 
> ```js
> https://next-dynamic-munnyreol.vercel.app/api/screenshot?url=https%3A%2F%2Fnextjs.org
> ```
> 
> ```js
> export default function handler(req, res) {
>   if(!req.param.url || Array.isArray(req.param.url)) {
>     res.status(400).end();
>     return;
>   }
>   
>   res.send(req.param.url);
> }
> ```

This is not working.

& There is no param object in req in nextjs.
That's why I replace param to query.
Still I'm getting same error.

Please check yourself, is this work or not. & Then .

**Comment by ljosberinn** [comment]
Sorry it's indeed `query`, but it works as expected (I'm using it in multiple sites).

```js
export default function handler(req, res) {
  console.log(req);
  res.send(JSON.stringify(req.query));
}
```

![image](https://user-images.githubusercontent.com/29307652/118613819-666d2e80-b7bf-11eb-98d0-590181076b38.png)

**Comment by MunnyReol** [comment]
> Sorry it's indeed `query`, but it works as expected (I'm using it in multiple sites).
> 
> ```js
> export default function handler(req, res) {
>   console.log(req);
>   res.send(JSON.stringify(req.query));
> }
> ```
> 
> ![image](https://user-images.githubusercontent.com/29307652/118613819-666d2e80-b7bf-11eb-98d0-590181076b38.png)


You are using `?` in url for parameters. That's why you can..
Your url is like `https://something.com/api/test?url=https%3A%2F%2Fnextjs.org`

But I'm not using `?` in url bar for parameters.
I'm using nextjs dynamic route like `[id].js` for parameter

That's why I'm getting error.

You can check my testing repository 
[https://github.com/MunnyReol/next-dynamic](https://github.com/MunnyReol/next-dynamic)

**Comment by ljosberinn** [comment]
Yes, as I said:

> This is not Next.js related but with how URLs work. Pass the url instead as GET parameter.

**Comment by MunnyReol** [comment]
> Yes, as I said:
> 
> > This is not Next.js related but with how URLs work. Pass the url instead as GET parameter.

Yes, you said.
But this is not the solve of my problem.
Because i don't use `?` in url, i need to to create a dynamic page like `[url].js`.
😌😌

**Comment by whoanuragverma** [comment]
You're facing this issue because you're actually using wrong file paths for dynamic api paths, it is clearly mentioned in [next.js docs](https://nextjs.org/docs/api-routes/dynamic-api-routes) about dynamic API routes. 

To put simply instead of the path as `pages/api/screenshot/[id]/index.js` make it as `pages/api/screenshot/[id].js` and use your own code you'll get desired response. Moreover, if you want to make nested links like `/api/screenshot/abc/def/cde` you could simply rename the file as `pages/api/screenshot/[...id].js`. 

Everything is mentioned pretty clearly in docs.

**Comment by MunnyReol** [comment]
> You're facing this issue because you're actually using wrong file paths for dynamic api paths, it is clearly mentioned in [next.js docs](https://nextjs.org/docs/api-routes/dynamic-api-routes) about dynamic API routes.
> 
> To put simply instead of the path as `pages/api/screenshot/[id]/index.js` make it as `pages/api/screenshot/[id].js` and use your own code you'll get desired response. Moreover, if you want to make nested links like `/api/screenshot/abc/def/cde` you could simply rename the file as `pages/api/screenshot/[...id].js`.
> 
> Everything is mentioned pretty clearly in docs.

Yes, you are right, but It's give me error.
Check my repository

[https://github.com/MunnyReol/next-dynamic](https://github.com/MunnyReol/next-dynamic) 

This is problem I'm fetching after deploy on vercel.
But in my local system, there is no error.

I think this is the problem because of vercel
